## Gather

In [1]:
import pandas as pd
import numpy as np

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

## Assess

In [3]:
patients

patient_id assigned_sex  given_name       surname  \
0             1       female         Zoe       Wellish   
1             2       female      Pamela          Hill   
2             3         male         Jae        Debord   
3             4         male        Liêm          Phan   
4             5         male         Tim       Neudorf   
5             6         male      Rafael         Costa   
6             7       female        Mary         Adams   
7             8       female      Xiuxiu         Chang   
8             9         male       Dsvid    Gustafsson   
9            10       female      Sophie       Cabrera   
10           11       female       Sandy    Gunnarsson   
11           12         male   Abdul-Nur           Isa   
12           13         male  Omeokachie      Ibeamaka   
13           14       female    Anenechi         Chidi   
14           15       female        Asia       Woźniak   
15           16         male       Søren          Lund   
16           17       female         Tám          Liễu   
17           18       female     Roxanne     Andreyeva   
18           19         male     William         Oates   
19           20         male         Zak         Kelly   
20           21       female       Sofia       Karlsen   
21           22         male      Samúel  Guðbrandsson   
22           23         male      Manchu            Su   
23           24         male       Lovre         Galić   
24           25         male       Jakob      Jakobsen   
25           26         male      Gregor          Bole   
26           27       female        Ella          Lund   
27           28         male      Joseph        Tucker   
28           29         male      Robert          Wolf   
29           30         male        Jake      Jakobsen   
..          ...          ...         ...           ...   
473         474       female        Kate     Wilkinson   
474         475       female   Esperanza      Labrosse   
475         476         male       Malik       Vaneker   
476         477       female       Berta    Napolitani   
477         478         male     Juliusz      Majewski   
478         479       female      Edelma   Villalpando   
479         480         male        Tapa   Arsanukayev   
480         481         male      Nasser       Mansour   
481         482         male     Michael    Kristensen   
482         483         male       Diogo         Souza   
483         484       female       Angel         Grant   
484         485         male     Placido       Udinesi   
485         486         male      Trifon      Izmailov   
486         487         male      Samuel          Blix   
487         488         male        Ivar       Löfgren   
488         489         male        Mika    Martinsson   
489         490       female     Jasmine         Sykes   
490         491         male     Jackson       Addison   
491         492       female     Vanessa      Ferguson   
492         493         male       Poldi           Tar   
493         494       female         Fen          Chin   
494         495       female      Sirkka     Piirainen   
495         496         male      Hajime       Tsukada   
496         497         male   Alexander        Hueber   
497         498         male    Masataka      Murakami   
498         499         male     Mustafa     Lindström   
499         500         male       Ruman       Bisliev   
500         501       female       Jinke     de Keizer   
501         502       female     Chidalu   Onyekaozulu   
502         503         male         Pat       Gersten   

                        address               city       state  zip_code  \
0          576 Brown Bear Drive  Rancho California  California   92390.0   
1     2370 University Hill Road          Armstrong    Illinois   61812.0   
2         1493 Poling Farm Road               York    Nebraska   68467.0   
3           2335 Webster Street         Woodbridge          NJ    7095.0   
4          1428 Turkey Pen L

In [4]:
treatments

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0       veronika      jindrová  41u - 48u          -         7.63       7.20   
1         elliot    richardson          -  40u - 45u         7.56       7.09   
2       yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3           skye    gormanston  33u - 36u          -         7.97       7.62   
4         alissa        montez          -  33u - 29u         7.78       7.46   
5        jasmine         sykes          -  42u - 44u         7.56       7.18   
6         sophia        haugen  37u - 42u          -         7.65       7.27   
7          eddie        archer  31u - 38u          -         7.89       7.55   
8          saber        ménard          -  54u - 54u         8.08       7.70   
9           asia       woźniak  30u - 36u          -         7.76       7.37   
10        joseph           day  29u - 36u          -         7.70       7.19   
11     kristiina        hyypiä          -  36u - 38u         7.87       7.49   
12       roxanne     andreyeva  29u - 38u          -         9.54       9.14   
13        gregor          bole          -  47u - 45u         7.61       7.16   
14        simone  baumgaertner  27u - 37u          -         7.74       7.30   
15          enco        žibrik  55u - 68u          -         7.78       7.34   
16       camilla      zaitseva  28u - 37u          -         7.53       7.13   
17          gina          cain          -  36u - 36u         7.88       7.40   
18    addolorata      lombardi          -  49u - 46u         7.75       7.33   
19        khalid      johnsrud          -  54u - 54u         8.35       7.94   
20          mile        stanić          -  47u - 48u         7.66       7.24   
21         tekla       walczak  29u - 39u          -         7.61       7.29   
22   brancaleone         russo  53u - 60u          -         8.61       8.18   
23      chiemela     tobeolisa          -  43u - 47u         7.59       7.17   
24          isac          berg  31u - 41u          -         9.68       9.29   
25        benoît        bonami          -  44u - 43u         9.82       9.40   
26        suhaim         rahal          -  49u - 47u         7.94       7.50   
27        mizuki         iwata          -  45u - 46u         7.70       7.23   
28       clinton        miller  42u - 51u          -         7.79       7.40   
29        eugene       mironov  42u - 49u          -         7.81       7.48   
..           ...           ...        ...        ...          ...        ...   
250         chen           yao          -  56u - 57u         7.90       7.51   
251        aksel   vestergaard          -  42u - 38u         9.62       9.29   
252        ellen         luman          -  40u - 39u         9.27       8.77   
253       albino     schiavone  35u - 43u          -         7.56       7.15   
254         jose         combs          -  39u - 36u         7.89       7.42   
255       jia li          teng  48u - 54u          -         7.66       7.32   
256        ilija        horvat  42u - 50u          -         7.77       7.38   
257     mathilde      nørgaard          -  27u - 28u         8.50       8.10   
258       csilla      herczegh          -  43u - 46u         7.71       7.27   
259      aaliyah          rice          -  31u - 31u         7.64       7.33   
260        david      beauvais          -  26u - 23u         7.87       7.47   
261     caroline        shuler          -  50u - 54u         7.63       7.27   
262         alex      crawford  51u - 62u          -         7.69       7.30   
263      rebecca      jephcott  53u - 63u          -         7.96       7.57   
264   chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
265       fearne      mcgregor          -  27u - 29u         7.83       7.48   
266       ursula         freud  42u - 54u          -         7.75       7.46   
267         leon        scholz          -  38u - 32u         7.72       7.29   
268   

In [5]:
adverse_reactions

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [6]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [7]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [8]:
adverse_reactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [9]:
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [10]:
list(patients)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'contact',
 'birthdate',
 'weight',
 'height',
 'bmi']

In [11]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [12]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [13]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [14]:
patients.sample(5)

patient_id assigned_sex  given_name    surname                address  \
86           87         male    Philemon    Abdulov       1341 Chatham Way   
402         403       female     Manouck    Wubbels  4789 Devils Hill Road   
364         365       female  Chukwumoge  Ogochukwu        4704 Edsel Road   
107         108         male       Rocco   Christie  1373 Wilmar Farm Road   
123         124         male         Tas      Fejes  4386 Badger Pond Lane   

             city state  zip_code        country  \
86   Temple Hills    MD   20031.0  United States   
402         Utica    MS   39175.0  United States   
364  Sherman Oaks    CA   91403.0  United States   
107    Mount Airy    MD   21771.0  United States   
123         Tampa    FL   33634.0  United States   

                                         contact  birthdate  weight  height  \
86         240-695-5212PhilemonAbdulov@rhyta.com   7/7/1943   171.4      70   
402       601-885-6550ManouckWubbels@armyspy.com  11/3/1964   201.5      65   
364  818-372-7106ChukwumogeOgochukwu@teleworm.us  9/16/1960   212.3      75   
107          240-322-1398RoccoChristie@rhyta.com  1/16/1987   167.0      70   
123          727-331-8429FejesTas@jourrapide.com  12/2/1982   193.2      66   

      bmi  
86   24.6  
402  33.5  
364  26.5  
107  24.0  
123  31.2

In [15]:
patients.surname.value_counts()

Doe            6
Taylor         3
Jakobsen       3
Lâm            2
Silva          2
Grímsdóttir    2
Dratchev       2
Cindrić        2
Batukayev      2
Hueber         2
Parker         2
Kadyrov        2
Ogochukwu      2
Gersten        2
Johnson        2
Kowalczyk      2
Berg           2
Schiavone      2
Woźniak        2
Lund           2
Correia        2
Liễu           2
Aranda         2
Collins        2
Bùi            2
Souza          2
Nilsen         2
Tucker         2
Lương          2
Cabrera        2
              ..
Tuma           1
Tobeolisa      1
Novosel        1
Totth          1
Kos            1
McGregor       1
Madrid         1
Mancini        1
Shuler         1
Klobučar       1
Galić          1
Lončar         1
Moore          1
Vaneker        1
Uspenskaya     1
Hsu            1
Komavec        1
Gyenes         1
Pecinová       1
Sleiman        1
Isa            1
Mortensen      1
Lê             1
Afanasyeva     1
Hill           1
Vaara          1
Reilly         1
Yonatan       

In [16]:
patients.address.value_counts()

123 Main Street             6
2778 North Avenue           2
648 Old Dear Lane           2
2476 Fulton Street          2
570 Alpha Avenue            1
3072 Braxton Street         1
4851 Andy Street            1
441 Tibbs Avenue            1
3499 Baker Avenue           1
192 Patton Lane             1
649 Nutter Street           1
2645 Moore Avenue           1
1079 Ingram Street          1
1330 Lincoln Street         1
2931 Romano Street          1
1731 Chandler Drive         1
3542 Robinson Court         1
3008 Walkers Ridge Way      1
494 Clarksburg Park Road    1
3113 Timber Ridge Road      1
475 Preston Street          1
4649 Worley Avenue          1
2578 Tenmile                1
1774 George Avenue          1
4310 Johnson Street         1
353 Whaley Lane             1
4111 Thunder Road           1
4386 Camden Street          1
4500 Myra Street            1
783 Callison Lane           1
                           ..
4839 North Avenue           1
577 Chipmunk Lane           1
2356 Myra 

In [17]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [18]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
171    106.5
51     107.1
270    108.1
198    108.5
48     109.1
478    109.6
141    110.2
38     111.8
438    112.0
14     112.0
235    112.2
307    112.4
191    112.6
408    113.1
49     113.3
326    114.0
338    114.1
253    117.0
321    118.4
168    118.8
1      118.8
350    119.0
207    119.2
265    120.0
341    120.3
       ...  
332    224.0
252    224.2
12     224.2
222    224.8
166    225.3
111    225.9
101    226.2
150    226.6
352    227.7
428    227.7
88     227.7
13     228.4
339    229.0
182    230.3
121    230.8
257    231.7
395    231.9
246    232.1
219    237.8
11     238.7
50     238.9
441    239.1
499    239.6
439    242.0
487    242.4
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [19]:
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / (height_in * height_in)
bmi_check

210    19.055827
dtype: float64

In [20]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

#### Quality
##### `patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) ***(can't clean yet)***
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

##### `treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

##### `adverse_reactions` table
- Lowercase given names and surnames

#### Tidiness
- Contact column in `patients` table should be split into phone number and email
- Three variables in two columns in `treatments` table (treatment, start dose and end dose)
- Adverse reaction should be part of the `treatments` table
- Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables

## Clean

In [23]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

### Missing Data

#### `treatments`: Missing records (280 instead of 350)

##### Define
Import the cut treatments into a DataFrame and concatenate it with the original treatments DataFrame.

##### Code

In [24]:
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_clean = pd.concat([treatments_clean, treatments_cut],
                             ignore_index=True)

##### Test

In [25]:
treatments_clean.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [26]:
treatments_clean.tail()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN

#### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

##### Define
Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`. 

##### Code

In [27]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

##### Test

In [28]:
treatments_clean.hba1c_change.head()

0    0.43
1    0.47
2    0.43
3    0.35
4    0.32
Name: hba1c_change, dtype: float64

### Tidiness

#### Contact column in `patients` table contains two variables: phone number and email

##### Define
Extract the *phone number* and *email* variables from the *contact* column using regular expressions and pandas' `str.extract` method. Drop the *contact* column when done.

##### Code

In [29]:
patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.][a-zA-Z]+)', expand=True)
# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

##### Test

In [30]:
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [31]:
patients_clean.phone_number.sample(25)

420         631-479-8171
278                  NaN
129         631-370-7406
201         603 773 2333
443         352-362-5392
479         909 458 2515
482         361-693-4960
461         646-289-4177
58          412-640-7035
501         360 443 2060
6           775-533-5933
291         302-698-2057
428         412-319-0903
183         909-355-9418
38          978 460 9060
322         254-546-2728
307         228-237-2271
151         916-224-7868
26          309-671-8852
161         406-759-6160
160         602-993-7880
283    +1 (708) 845-2053
92          609-914-8473
446         601-418-0102
419         317-956-6166
Name: phone_number, dtype: object

In [32]:
# Confirm that no emails start with an integer (regex didn't match for this)
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object

#### Three variables in two columns in `treatments` table (treatment, start dose and end dose)

##### Define
Melt the *auralin* and *novodra* columns to a *treatment* and a *dose* column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain *start_dose* and *end_dose* columns. Drop the intermediate *dose* column.

##### Code

In [33]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')
treatments_clean = treatments_clean[treatments_clean.dose != "-"]
treatments_clean['dose_start'], treatments_clean['dose_end'] = treatments_clean['dose'].str.split(' - ', 1).str
treatments_clean = treatments_clean.drop('dose', axis=1)

##### Test

In [34]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  dose_start dose_end  
0        41u      48u  
3        33u      36u  
6        37u      42u  
7        31u      38u  
9        30u      36u

#### Adverse reaction should be part of the `treatments` table

##### Define
Merge the *adverse_reaction* column to the `treatments` table, joining on *given name* and *surname*.

##### Code

In [35]:
treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean,
                            on=['given_name', 'surname'], how='left')

##### Test

In [36]:
treatments_clean

given_name        surname  hba1c_start  hba1c_end  hba1c_change  \
0        veronika       jindrová         7.63       7.20          0.43   
1            skye     gormanston         7.97       7.62          0.35   
2          sophia         haugen         7.65       7.27          0.38   
3           eddie         archer         7.89       7.55          0.34   
4            asia        woźniak         7.76       7.37          0.39   
5          joseph            day         7.70       7.19          0.51   
6         roxanne      andreyeva         9.54       9.14          0.40   
7          simone   baumgaertner         7.74       7.30          0.44   
8            enco         žibrik         7.78       7.34          0.44   
9         camilla       zaitseva         7.53       7.13          0.40   
10          tekla        walczak         7.61       7.29          0.32   
11    brancaleone          russo         8.61       8.18          0.43   
12           isac           berg         9.68       9.29          0.39   
13        clinton         miller         7.79       7.40          0.39   
14         eugene        mironov         7.81       7.48          0.33   
15    szilveszter          totth         7.70       7.38          0.32   
16      alexander      mathiesen         7.96       7.55          0.41   
17            chỉ            lâm         7.68       7.24          0.44   
18      władysław      wieczorek         7.92       7.47          0.45   
19       kristján        ingason         7.92       7.57          0.35   
20         marija       grubišić         7.53       7.15          0.38   
21          sauli     koivuniemi         7.67       7.37          0.30   
22        mariana          souza         7.86       7.51          0.35   
23     kristoffer      martinsen         9.18       8.64          0.54   
24             mỹ          quynh         7.61       7.16          0.45   
25           oles        zhdanov         7.52       7.11          0.41   
26        triana.       terrazas         7.71       7.34          0.37   
27         gabryŝ    tomaszewski         7.87       7.47          0.40   
28      leixandre         alanis         7.74       7.32          0.42   
29   onyekachukwu         obinna         7.58       7.12          0.46   
..            ...            ...          ...        ...           ...   
320          jane        citizen         7.98       7.60          0.38   
321        angela     lavrentyev         7.61       7.14          0.47   
322        edelma    villalpando         7.99       7.56          0.43   
323        annika          vaara         7.73       7.34          0.39   
324         chiho           higa         7.71       7.30          0.41   
325     beatrycze        woźniak         7.54       7.17          0.37   
326      miłosław     wiśniewski         7.51       7.08          0.43   
327       firenze          fodor         7.89       7.55          0.34   
328           zoe        wellish         7.71       7.30          0.41   
329           una  traustadóttir         8.00       7.50          0.50   
330         luboš          pecha         7.79       7.45          0.34   
331         meaza         brhane         7.70       7.36          0.34   
332         adlan       shishani         7.84       7.37          0.47   
333         sofia      hermansen         8.90       8.57          0.33   
334         guðni      heimisson         7.64       7.24          0.40   
335       eufemio        rosario         7.54       7.26          0.28   
336      dalmacia         madrid         7.67       7.21          0.46   
337         daimy          tromp         9.41       8.94          0.47   
338        jeremy        montagu         7.68       7.36          0.32   
339       nebechi      ekechukwu         7.78       7.39          0.39   
340       satsita      batukayev         7.63       7.25          0.38   
341       timothy         cotton         7.92       7.52          0.40   
342       b

#### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables  and Lowercase given names and surnames

##### Define
Adverse reactions table is no longer needed so ignore that part. Isolate the patient ID and names in the `patients` table, then convert these names to lower case to join with `treatments`. Then drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).

##### Code

In [37]:
id_names = patients_clean[['patient_id', 'given_name', 'surname']]
id_names.given_name = id_names.given_name.str.lower()
id_names.surname = id_names.surname.str.lower()
treatments_clean = pd.merge(treatments_clean, id_names, on=['given_name', 'surname'])
treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

/opt/conda/lib/python3.6/site-packages/pandas/core/generic.py:4405: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[name] = value


##### Test

In [38]:
# Confirm the merge was executed correctly
treatments_clean

hba1c_start  hba1c_end  hba1c_change treatment dose_start dose_end  \
0           7.63       7.20          0.43   auralin        41u      48u   
1           7.97       7.62          0.35   auralin        33u      36u   
2           7.65       7.27          0.38   auralin        37u      42u   
3           7.89       7.55          0.34   auralin        31u      38u   
4           7.76       7.37          0.39   auralin        30u      36u   
5           7.70       7.19          0.51   auralin        29u      36u   
6           7.70       7.19          0.51   auralin        29u      36u   
7           9.54       9.14          0.40   auralin        29u      38u   
8           7.74       7.30          0.44   auralin        27u      37u   
9           7.78       7.34          0.44   auralin        55u      68u   
10          7.53       7.13          0.40   auralin        28u      37u   
11          7.61       7.29          0.32   auralin        29u      39u   
12          8.61       8.18          0.43   auralin        53u      60u   
13          9.68       9.29          0.39   auralin        31u      41u   
14          7.79       7.40          0.39   auralin        42u      51u   
15          7.81       7.48          0.33   auralin        42u      49u   
16          7.70       7.38          0.32   auralin        35u      39u   
17          7.96       7.55          0.41   auralin        47u      58u   
18          7.68       7.24          0.44   auralin        45u      48u   
19          7.92       7.47          0.45   auralin        24u      37u   
20          7.92       7.57          0.35   auralin        44u      55u   
21          7.53       7.15          0.38   auralin        37u      43u   
22          7.67       7.37          0.30   auralin        43u      47u   
23          7.86       7.51          0.35   auralin        36u      42u   
24          9.18       8.64          0.54   auralin        29u      37u   
25          7.61       7.16          0.45   auralin        57u      64u   
26          7.52       7.11          0.41   auralin        54u      67u   
27          7.71       7.34          0.37   auralin        34u      42u   
28          7.87       7.47          0.40   auralin        29u      37u   
29          7.74       7.32          0.42   auralin        61u      67u   
..           ...        ...           ...       ...        ...      ...   
319         7.98       7.60          0.38   novodra        37u      38u   
320         7.61       7.14          0.47   novodra        28u      24u   
321         7.99       7.56          0.43   novodra        24u      26u   
322         7.73       7.34          0.39   novodra        20u      21u   
323         7.71       7.30          0.41   novodra        46u      46u   
324         7.54       7.17          0.37   novodra        26u      27u   
325         7.51       7.08          0.43   novodra        34u      33u   
326         7.89       7.55          0.34   novodra        30u      35u   
327         7.71       7.30          0.41   novodra        33u      33u   
328         8.00       7.50          0.50   novodra        35u      34u   
329         7.79       7.45          0.34   novodra        30u      27u   
330         7.70       7.36          0.34   novodra        37u      41u   
331         7.84       7.37          0.47   novodra        43u      40u   
332         8.90       8.57          0.33   novodra        34u      34u   
333         7.64       7.24          0.40   novodra        40u      36u   
334         7.54       7.26          0.28   novodra        37u      40u   
335         7.67       7.21          0.46   novodra        26u      23u   
336         9.41       8.94          0.47   novodra        40u      45u   
337         7.68       7.36          0.32   novodra        52u      52u   
338         7.78       7.39          0.39   novodra        37u      39u   
339         7.63       7.25          0.38   novodra        42u      42u   
340         7.92       7.52   

In [39]:
# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality

#### Zip code is a float not a string and Zip code has four digits sometimes

##### Define
Convert the zip code column's data type from a float to a string using `astype`, remove the '.0' using string slicing, and pad four digit zip codes with a leading 0.

##### Code

In [40]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')
# Reconvert NaNs entries that were converted to '0000n' by code above
patients_clean.zip_code = patients_clean.zip_code.replace('0000n', np.nan)

##### Test

In [41]:
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

#### Tim Neudorf height is 27 in instead of 72 in

##### Define
Replace height for rows in the `patients` table that have a height of 27 in (there is only one) with 72 in.

##### Code

In [42]:
patients_clean.height = patients_clean.height.replace(27, 72)

##### Test

In [43]:
# Should be empty
patients_clean[patients_clean.height == 27]

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [44]:
# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf']

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      72  26.1   

   phone_number                email  
4  334-515-7487  TimNeudorf@cuvox.de

#### Full state names sometimes, abbreviations other times

##### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.

##### Code

In [45]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

##### Test

In [46]:
patients_clean.state.value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
LA    13
MI    13
OK    13
NJ    12
VA    11
WI    10
MS    10
AL     9
MN     9
TN     9
IN     9
WA     8
NC     8
KY     8
MO     7
KS     6
NV     6
ID     6
NE     6
CT     5
SC     5
IA     5
AR     4
RI     4
ME     4
AZ     4
CO     4
ND     4
OR     3
DE     3
WV     3
SD     3
MD     3
MT     2
VT     2
DC     2
AK     1
NH     1
WY     1
NM     1
Name: state, dtype: int64

In [64]:
patients_clean.state

0      CA
1      IL
2      NE
3      NJ
4      AL
5      FL
6      NV
7      CA
8      MO
9      NY
10     MI
11     TN
12     VA
13     AL
14     OK
15     VA
16     CA
17     GA
18     MT
19     CA
20     MA
21     NJ
22     NY
23     FL
24     NY
25     NM
26     IL
27     LA
28     PA
30     CO
       ..
472    MS
473    MA
474    GA
475    CA
476    PA
477    SC
478    CA
479    CA
480    TX
481    OK
482    TX
483    TN
484    NC
485    FL
486    TX
487    KS
488    FL
489    CA
490    NY
491    TX
492    CA
493    ID
494    MI
495    CA
496    CA
497    OK
498    ME
499    AZ
500    MO
501    WA
Name: state, Length: 494, dtype: category
Categories (49, object): [AK, AL, AR, AZ, ..., WA, WI, WV, WY]

#### Dsvid Gustafsson

##### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

##### Code

In [47]:
patients_clean.given_name = patients_clean.given_name.replace('Dsvid', 'David')

##### Test

In [48]:
patients_clean[patients_clean.surname == 'Gustafsson']

patient_id assigned_sex given_name     surname             address  \
8           9         male      David  Gustafsson  1790 Nutter Street   

          city state zip_code        country birthdate  weight  height   bmi  \
8  Kansas City    MO    64105  United States  3/6/1937   163.9      66  26.5   

   phone_number                        email  
8  816-265-9578  DavidGustafsson@armyspy.com

#### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra

##### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

##### Code

In [49]:
# To category
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

# To datetime
patients_clean.birthdate = pd.to_datetime(patients_clean.birthdate)

# Strip u and to integer
treatments_clean.dose_start = treatments_clean.dose_start.str.strip('u').astype(int)
treatments_clean.dose_end = treatments_clean.dose_end.str.strip('u').astype(int)

##### Test

In [50]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null category
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null category
zip_code        491 non-null object
country         491 non-null object
birthdate       503 non-null datetime64[ns]
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
phone_number    491 non-null object
email           491 non-null object
dtypes: category(2), datetime64[ns](1), float64(2), int64(2), object(8)
memory usage: 53.9+ KB


In [51]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 348
Data columns (total 8 columns):
hba1c_start         349 non-null float64
hba1c_end           349 non-null float64
hba1c_change        349 non-null float64
treatment           349 non-null object
dose_start          349 non-null int64
dose_end            349 non-null int64
adverse_reaction    35 non-null object
patient_id          349 non-null int64
dtypes: float64(3), int64(3), object(2)
memory usage: 24.5+ KB


#### Multiple phone number formats

##### Define
Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code).

##### Code

In [52]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '').str.pad(11, fillchar='1')

##### Test

In [53]:
patients_clean.phone_number.head()

0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

#### Default John Doe data

##### Define
Remove the non-recoverable John Doe records from the `patients` table.

##### Code

In [54]:
patients_clean = patients_clean[patients_clean.surname != 'Doe']

##### Test

In [55]:
# Should be no Doe records
patients_clean.surname.value_counts()

Jakobsen       3
Taylor         3
Dratchev       2
Berg           2
Kowalczyk      2
Aranda         2
Grímsdóttir    2
Gersten        2
Silva          2
Johnson        2
Lâm            2
Cabrera        2
Ogochukwu      2
Collins        2
Cindrić        2
Batukayev      2
Hueber         2
Kadyrov        2
Nilsen         2
Parker         2
Schiavone      2
Liễu           2
Tạ             2
Woźniak        2
Souza          2
Tucker         2
Lund           2
Bùi            2
Lương          2
Correia        2
              ..
Isaksson       1
Tobeolisa      1
Novosel        1
Totth          1
Kos            1
McGregor       1
Madrid         1
Mancini        1
Shuler         1
Klobučar       1
Galić          1
Lončar         1
Moore          1
Vaneker        1
Uspenskaya     1
Hsu            1
Komavec        1
Gyenes         1
Pecinová       1
Sleiman        1
Isa            1
Mortensen      1
Lê             1
Afanasyeva     1
Hill           1
Vaara          1
Reilly         1
Yonatan       

In [56]:
# Should be no 123 Main Street records
patients_clean.address.value_counts()

2778 North Avenue           2
648 Old Dear Lane           2
2476 Fulton Street          2
3542 Robinson Court         1
3499 Baker Avenue           1
1079 Ingram Street          1
3072 Braxton Street         1
570 Alpha Avenue            1
4851 Andy Street            1
441 Tibbs Avenue            1
192 Patton Lane             1
649 Nutter Street           1
2645 Moore Avenue           1
1330 Lincoln Street         1
2931 Romano Street          1
1731 Chandler Drive         1
3008 Walkers Ridge Way      1
494 Clarksburg Park Road    1
3113 Timber Ridge Road      1
475 Preston Street          1
4649 Worley Avenue          1
2578 Tenmile                1
1774 George Avenue          1
4310 Johnson Street         1
353 Whaley Lane             1
4111 Thunder Road           1
4386 Camden Street          1
4500 Myra Street            1
783 Callison Lane           1
3464 Big Indian             1
                           ..
4839 North Avenue           1
577 Chipmunk Lane           1
2356 Myra 

#### Multiple records for Jakobsen, Gersten, Taylor

##### Define
Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the `patients` table. These are the nicknames, which happen to also not be in the `treatments` table (removing the wrong name would create a consistency issue between the `patients` and `treatments` table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.

##### Code

In [57]:
# tilde means not: http://pandas.pydata.org/pandas-docs/stable/indexing.html#boolean-indexing
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]

##### Test

In [58]:
patients_clean[patients_clean.surname == 'Jakobsen']

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [59]:
patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [60]:
patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

#### kgs instead of lbs for Zaitseva weight

##### Define
Use [advanced indexing](https://stackoverflow.com/a/44913631) to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

##### Code

In [61]:
weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

##### Test

In [62]:
# 48.8 shouldn't be the lowest anymore
patients_clean.weight.sort_values()

459    102.100000
335    102.700000
74     103.200000
317    106.000000
171    106.500000
51     107.100000
210    107.585456
270    108.100000
198    108.500000
48     109.100000
478    109.600000
141    110.200000
38     111.800000
438    112.000000
14     112.000000
235    112.200000
307    112.400000
191    112.600000
408    113.100000
49     113.300000
326    114.000000
338    114.100000
253    117.000000
321    118.400000
168    118.800000
1      118.800000
350    119.000000
207    119.200000
265    120.000000
341    120.300000
          ...    
332    224.000000
12     224.200000
252    224.200000
222    224.800000
166    225.300000
111    225.900000
101    226.200000
150    226.600000
88     227.700000
352    227.700000
428    227.700000
13     228.400000
339    229.000000
182    230.300000
121    230.800000
257    231.700000
395    231.900000
246    232.100000
219    237.800000
11     238.700000
50     238.900000
441    239.100000
499    239.600000
439    242.000000
487    242